### ISSUE TITLE
Straighten out handling of numeric and boolean literals

### ISSUE THREAD

**Issue by ojwb** [swig/swig#2995]
While working on #2990, I noticed that multiple backends have grown their own machinery to try to turn a C/C++ numeric or boolean literal into one suitable for use in the target language.  E.g.:

* C#: `true` -> `1`; other `T_BOOL` -> `0` (which is buggy as it means `typedef enum { T = (bool)1 } b;` currently generates `T = 0` in C#!)
* D: `true` -> `1`; other `T_BOOL` -> `0` (which is buggy as it means `typedef enum { T = (bool)1 } b;` currently generates `T = 0` in D!)
* Go: `true` -> `1`; `false` -> `0`; "Accept a 0x prefix, and strip combinations of u and l suffixes"
* Java: `true` -> `1`; other `T_BOOL` -> `0` (which is buggy as it means `typedef enum { T = (bool)1 } b;` currently generates `T = new p("T", 0)` in Java with `%javaconst(1);`!)
* Octave: `true`/`TRUE` -> `true`; `false`/`FALSE` -> `false`; value which starts with `[0-9]` seems to be used as an Octave literal which seems buggy but I've not tested; (also `NULL`/`nullptr` with pointer type -> `nil`, while `0` with pointer type -> `None` which seems odd - it looks like this code was probably based on the Ruby version, so `nil` may just be wrong as at least based on our examples and testcases, `None` is correct for Octave)
* Python: `true` -> `True`; `false` -> `False`; removes any `[lLuU]\+$` suffix (if `[uU]` present then negative values not allowed); integer value with boolean type -> `True` or `False`; integer value `0` for pointer type -> `None`; attempts to handle octal and hex integer literals too; removes `[fF]+$` suffix from double literals; (also `NULL`/`nullptr` with pointer type -> `None`, but with non-pointer type -> `0`)
* R: `true` -> `1`; other `T_BOOL` -> `0` (which is presumably buggy like in D and Java)
* Ruby: `true`/`TRUE` -> `True`; `false`/`FALSE` -> `False`; value which starts with `[0-9]` seems to be used as a Ruby literal which seems buggy but I've not tested; (also `NULL`/`nullptr` with pointer type -> `nil`, but with non-pointer type -> `0`)

The PHP backend used to have something like this too, but I removed it in 4.1.0 as we no longer generate any PHP code (everything is done in the C/C++ wrapper now).

There's a lot of duplication of functionality here (and of code too, as Octave seems to have copied Ruby, while D and C# look to be based on Java), and I noticed quite a few bugs just from a brief scan of the code - if we had a single implementation it would get a lot more testing than 8 separate implementations will).  It's also unnecessarily inconsistent what is supported across different target languages, which is unhelpful to users working with more than one target language.

I think this would be better done by having the parser add an `numval` or similar member which contains the integer value as a simple string of ASCII digits (and is absent when that's not possible), and a backend wants to generate a target language integer literal it can just use this (with any adjustments specific to the target language).  That would bring all target languages that want to do this up to a common level, and future improvements to this common machinery would automatically benefit all target languages (e.g. once #1094 gets addressed, all target languages could automatically handle C23/C++14 literals such as `1'000` too).

I think it makes sense to treat `bool` as an integer type here and normalise `false` to `0` and `true` to `1`.

It seems we should have some common handling for `NULL` and `nullptr` literals (and `0` in a pointer context too) since multiple target languages want it, so maybe those should set `numval=0`.

Floating point literals could be handled too, but seem a bit more complicated.  We'd need to decide what the normalised form is - e.g. should a value in exponential notation be converted to a string of digits with a decimal point?  Should the value always include a decimal point?  Should the decimal point always have a digit before and after it?  Probably needs a bit of research on what floating point literal formats are portable across target languages to determine what the lowest common denominator actually is.  Possibly the target language could tell the parser what it can handle if the lowest common denominator seems too low.

I'll probably wait for #2990 to get merged first, assuming that doesn't drag out.

Adding to the 4.3 milestone - the whole thing here isn't a release blocker, but if I don't get this done for 4.3 we should consider if we want to address the bugs such as `(bool)1` currently being treated as `false` by multiple target languages.

**Comment by ojwb** [comment]
I poked the Octave `nil` vs `None` case a bit but I can't work out how to trigger generation of either.  It looks like the Octave code is only for generating documentation comments (with `%feature("autodoc", 1);` on).

**Comment by ojwb** [comment]
I've done significant work on this.  I've still to look at Ruby, but the other languages seem OK except for an issue with handling default arguments in Python.  I managed to find a testcase in the testsuite which generates a doc comment with one of the cases in, so maybe I can manage to trigger the other there too.

With this and the previous refactor for string and character literals I ran into issues due to failing to initialise fields in the parser tokens.  It's made worse because we often get away with it in a simple case when the token is from fresh stack space, which will typically get zero initialised, but a reused stack slot can have arbitrary values in (but also they're often zero).  I suspect we have latent bugs due to this and I wonder if we should find a clean way to zero initialise tokens, either by default or explicitly.  I can't see an sort of token "constructor" in the bison docs (unless we switch to generating C++ code and then use class constructors), even though it does have a way to specify a token destructor.  So maybe something like:

```
def_args       : EQUAL definetype { 
                 $$ = $definetype;
               }
               | EQUAL definetype LBRACKET expr RBRACKET {                      
                 $$ = $definetype;
                 $$.val = NewStringf("%s[%s]", $definetype.val, $expr.val);   
               }                                                                 
               | EQUAL LBRACE {                                                 
                 if (skip_balanced('{','}') < 0) Exit(EXIT_FAILURE);
                 init_dtype($$);
                 $$.type = T_UNKNOWN;
                 $$.val = NewString(scanner_ccode);
               }
               | COLON expr {                                                 
                 init_dtype($$);
                 $$.bitfield = $expr.val;                                   
               }
               | %empty {
                 init_dtype($$);
                 $$.type = T_UNKNOWN;
               }
               ;
```

instead of the current code:

```
def_args       : EQUAL definetype {                    
                 $$ = $definetype;                  
               }
               | EQUAL definetype LBRACKET expr RBRACKET {
                 $$ = $definetype;
                 $$.val = NewStringf("%s[%s]", $definetype.val, $expr.val);
               }
               | EQUAL LBRACE { 
                 if (skip_balanced('{','}') < 0) Exit(EXIT_FAILURE);
                 $$.val = NewString(scanner_ccode);
                 $$.stringval = 0;   
                 $$.type = T_UNKNOWN;               
                 $$.unary_arg_type = 0;             
                 $$.bitfield = 0;
                 $$.throws = 0;
                 $$.throwf = 0;
                 $$.nexcept = 0;
                 $$.final = 0;
               }
               | COLON expr { 
                 $$.val = 0;
                 $$.stringval = 0;
                 $$.type = 0;
                 $$.bitfield = $expr.val;
                 $$.throws = 0;
                 $$.throwf = 0;
                 $$.nexcept = 0;
                 $$.final = 0;
               }
               | %empty {
                 $$.val = 0;
                 $$.stringval = 0; 
                 $$.type = T_UNKNOWN;
                 $$.unary_arg_type = 0;
                 $$.bitfield = 0;                
                 $$.throws = 0;
                 $$.throwf = 0;
                 $$.nexcept = 0;
                 $$.final = 0;                                               
               }
               ;
```

This seems both less error-prone and less verbose (which means you can more clearly see the parts that aren't just default initialisations of fields).

Note that here the first two rules don't need to `init_dtype($$);` because they assign to `$$` directly.

We could instead have something like:

```
   $$ = default_dtype;
```

where `default_dtype` is a `const static` instance, so the compiler would automatically initialise all the fields to zero for us.

**Comment by ojwb** [comment]
> Adding to the 4.3 milestone - the whole thing here isn't a release blocker, but if I don't get this done for 4.3 we should consider if we want to address the bugs such as `(bool)1` currently being treated as `false` by multiple target languages.

I'll see if I can get my branch merged for 4.3.0.  I don't think it's too far off.

**Comment by ojwb** [comment]
> R: `true` -> `1`; other `T_BOOL` -> `0` (which is presumably buggy like in D and Java)

It seems SWIG/R (incorrectly) updates the `enumvalue` attribute, but then never actually uses it so for SWIG/R this bug appears to be latent.

**Comment by ojwb** [comment]
> * Ruby: [...] value which starts with `[0-9]` seems to be used as a Ruby literal which seems buggy but I've not tested; [...]

In fact this doesn't make a difference because for cases where it might it only leads to returning the original value or `NULL`, and the only caller handles `NULL` by just using the original value!

**Comment by ojwb** [comment]
I've managed to write a regression test for D.  I can't seem manage to for C# or Java.  They generated wrong code which compiled before, but the expressions which trigger this aren't valid in C# or Java in this context, so the fixed version either fetches the enum value from C/C++ (and so doesn't test what we want) or it doesn't compile if we force it to use the C/C++ enumerator initialiser in C#/Java.

It's time I stopped for today, but will tidy this up and push.

Another point to resolve is whether to treat `TRUE` and `FALSE` the same as `true` and `false`, like some target languages currently do.  I think we shouldn't as they aren't standard C/C++, this assumption could break things (e.g. `#define TRUE -1` would be treated as `1`), and it's easy to `#define TRUE true` (or to `1`), etc in your interface file if you actually want this to happen.  This doesn't need to block merging this branch since it's an existing quirk, just one I came across while working on this.
